Cython/Build/Tests/TestRecythonize.py

```diff
@@ -0,0 +1,213 @@
+import shutil
+import os
+import tempfile
+
+import Cython.Build.Dependencies
+import Cython.Utils
+from Cython.TestUtils import CythonTest
+
+
+
+SAME = "The result of cytonization is the same"
+INCORRECT = "Incorrect cythonization"
+LINE_BEFORE_IMPLEMENTATION = '  /* "{filename}":{at_line}\n'
+VARS_LINE = '  /*--- Wrapped vars code ---*/\n'
+
+
+class TestRecythonize(CythonTest):
+    language_level = 3
+    dep_tree = Cython.Build.Dependencies.create_dependency_tree()
+
+    def сlear_function_and_Dependencies_caches(self):
+        Cython.Utils.clear_function_caches()
+        Cython.Build.Dependencies._dep_tree = None  # discard method caches
+
+    def setUp(self):
+        CythonTest.setUp(self)
+        self.сlear_function_and_Dependencies_caches()
+        self.temp_dir = (
+            tempfile.mkdtemp(
+                prefix='recythonize-test',
+                dir='TEST_TMP' if os.path.isdir('TEST_TMP') else None
+            )
+        )
+        self.src_dir = os.path.join(self.temp_dir, 'src')
+        os.mkdir(self.src_dir)  # self.temp_dir is already a temp
+
+    def tearDown(self):
+        CythonTest.tearDown(self)
+        self.сlear_function_and_Dependencies_caches()
+        shutil.rmtree(self.temp_dir)
+
+    def write_to_file(self, path, text):
+        with open(path, "w") as f:
+            f.write(text)
+
+    def fresh_cythonize(self, *args, **kwargs):
+        self.сlear_function_and_Dependencies_caches()
+        kwargs.update(language_level=self.language_level)
+        Cython.Build.Dependencies.cythonize(*args, **kwargs)
```

Comment:
I think it will be useful to move the auxiliary functions, such as [`fresh_cythonize`](https://github.com/cython/cython/pull/4241/files#diff-e4fc2a69422e9be305ea77db346411e8e2e3e179943ee3ef90cf1f00e64e7b7cR46-R49), [`write_to_file`](https://github.com/cython/cython/pull/4241/files#diff-e4fc2a69422e9be305ea77db346411e8e2e3e179943ee3ef90cf1f00e64e7b7cR42-R44) and [`сlear_function_and_Dependencies_caches`](https://github.com/cython/cython/pull/4241/files#diff-e4fc2a69422e9be305ea77db346411e8e2e3e179943ee3ef90cf1f00e64e7b7cR21-R23), into a separate file with the name, for example, `UnitTestHelpers`. Moreover, these functions are also used / can already be used with other files, for example [`TestCyCache.py`](https://github.com/cython/cython/blob/master/Cython/Build/Tests/TestCyCache.py)

Reply:
*A version of `fresh_cythonize` without `language_level`:
```python
def fresh_cythonize(self, *args, **kwargs):
    self.сlear_function_and_Dependencies_caches()
    Cython.Build.Dependencies.cythonize(*args, **kwargs)
```

Reply:
@scoder What do you think is worth moving these functions to another file?

Reply:
@scoder well, are we going to carry out the [same functionality](https://github.com/cython/cython/pull/4241#discussion_r657052375) outside the test cases?
It's just that otherwise it won't be very DRY to have the same `fresh_cythonize` in two different files.

Well also, [which name is better](https://github.com/cython/cython/pull/4241#discussion_r657046103)?

Reply:
@scoder I never got a single answer to my questions in [this](https://github.com/cython/cython/pull/4241#discussion_r660030772) conversation. If you think that it is not worth doing, say so, if you are thinking about it, then please say it.
I just don't understand why you are responding to other conversations but ignoring this one.

Reply:
Sorry, seems that I overlooked it. Certainly not intentional. There are just way too many comments on this PR, which makes it easy for me as a human being to miss some. Less is certainly more, when it comes to ticket conversations.

To answer your questions, feel free to move these three methods as functions into `Cython/TestUtils.py` (which, BTW, is a bad name because it starts with `Test`, suggesting a test module, but does not contain any tests… oh, well…). Their names look good as they stand.

Reply:
> Sorry, seems that I overlooked it. Certainly not intentional

We are people, it's okay to overlook something, but it's already good that you answered at least now :)

Reply:
> `Cython/TestUtils.py` (which, BTW, is a bad name because it starts with `Test`, suggesting a test module, but does not contain any tests… oh, well…)

This is not the best name, I see. Is it possible to fix this by calling it "UtilsForTests"/"TestingUtils"?

Reply:
> > `Cython/TestUtils.py` (which, BTW, is a bad name because it starts with `Test`, suggesting a test module, but does not contain any tests… oh, well…)
> 
> This is not the best name, I see. Is it possible to fix this by calling it "UtilsForTests"/"TestingUtils"?

1) it's not much of an issue, 2) not in this PR, 3) maybe, at some point, 4) not the second name, since that also starts with `Test`.

Reply:
> 1. it's not much of an issue, 2) not in this PR, 3) maybe, at some point

Looks to me, so it won't change for very long if it doesn't change now. But as you wish, as a matter of fact.

Reply:
partially resolved [here](https://github.com/cython/cython/pull/4241/commits/4e3f3c226271c024ddb3332dbdc279eb200d1d85), [here](https://github.com/cython/cython/pull/4241/commits/bbcf2dab4482f991adbf7625b91babac561c9c7a) and [here](https://github.com/cython/cython/pull/4241/commits/5162fa2dab618f600eaed82690d839ce67f7a30c)

Reply:
or let's resolve this conversation